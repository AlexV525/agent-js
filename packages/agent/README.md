# @dfinity/agent

JavaScript and TypeScript library to interact with the [Internet Computer](https://dfinity.org/) for Node.js and Client applications

Visit the [Dfinity Forum](https://forum.dfinity.org/) and [SDK Documentation](https://internetcomputer.org/docs/index.html) for more information and support building on the Internet Computer. For an introduction to the JavaScript / TypeScript agent, see [internetcomputer.org - JavaScript Intro](https://internetcomputer.org/docs/current/developer-docs/build/agents/javascript/javascript-intro).

Additional API Documentation can be found at [agent-js.icp.xyz](https://agent-js.icp.xyz/agent/index.html), and the changelog is available at [agent-js.icp.xyz/changelog.html](agent-js.icp.xyz/changelog.html) or alternately at [github releases](https://github.com/dfinity/agent-js/releases).

---

## Installation

Using agent:

```
npm i --save @dfinity/agent
```

### In the browser:

```
import * as agent from "@dfinity/agent";
```

or using individual exports:

```
import { Actor, HttpAgent } from "@dfinity/agent";
```

### In Node.js

```
const actor = require("@dfinity/agent");
```

Important Contents of this package:

- [Actor](https://agent-js.icp.xyz/agent/classes/Actor.html)
- [HttpAgent](https://agent-js.icp.xyz/agent/classes/HttpAgent.html)
- [Certificate](https://agent-js.icp.xyz/agent/classes/Certificate.html)
- [CanisterStatus](https://agent-js.icp.xyz/agent/modules/CanisterStatus.html)
- Utils
  - Buffer Utils
    - [compare](https://agent-js.icp.xyz/agent/modules.html#compare)
    - [concat](https://agent-js.icp.xyz/agent/modules.html#concat)
    - [fromHex](https://agent-js.icp.xyz/agent/modules.html#fromHex)
    - [hash](https://agent-js.icp.xyz/agent/modules.html#hash) - sha256 hash of an ArrayBuffer
    - [toHex](https://agent-js.icp.xyz/agent/modules.html#toHex)
  - IC Utils
    - [createAssetCanisterActor](https://agent-js.icp.xyz/agent/modules.html#createAssetCanisterActor)
    - [getDefaultAgent](https://agent-js.icp.xyz/agent/modules.html#getDefaultAgent)
    - [getManagementCanister](https://agent-js.icp.xyz/agent/modules.html#getManagementCanister)

## Using an Agent

The agent is a low-level interface that the Actor uses to encode and decode messages to the Internet Computer. It provides `call`, `query` and `readState` methods to the Actor, as well as a few additional utilities. For the most part, calls through the agent are intended to be structured through an Actor, configured with a canister interface that can be automatically generated from a [Candid](https://github.com/dfinity/candid) interface.

## Initializing an Actor

The most common use for the agent is to create an actor. This is done by calling the `Actor.createActor` constructor:

```
Actor.createActor(interfaceFactory: InterfaceFactory, configuration: ActorConfig): ActorSubclass<T>
```

The `interfaceFactory` is a function that returns a runtime interface that the Actor uses to strucure calls to a canister. The interfaceFactory can be written manually, but it is recommended to use the `dfx generate` command to generate the interface for your project, or to use the `didc` tool to generate the interface for your project.

### Inspecting an actor's agent

Use the `Actor.agentOf` method to get the agent of an actor:

```
const defaultAgent = Actor.agentOf(defaultActor);
```

This is useful if you need to replace or invalidate the identity used by an actor's agent.

For example, if you want to replace the identity of an actor's agent with a newly authenticated identity from [Internet Identity](https://identity.ic0.app), you can do so by calling the `Actor.replaceAgent` method:

```
defaultAgent.replaceIdentity(await authClient.getIdentity());
```
