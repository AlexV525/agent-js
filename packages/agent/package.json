{
  "name": "@dfinity/agent",
  "version": "0.15.7",
  "author": "DFINITY Stiftung <sdk@dfinity.org>",
  "license": "Apache-2.0",
  "description": "JavaScript and TypeScript library to interact with the Internet Computer",
  "homepage": "https://internetcomputer.org",
  "repository": {
    "type": "git",
    "url": "https://github.com/dfinity/agent-js.git",
    "directory": "packages/agent"
  },
  "bugs": {
    "url": "https://github.com/dfinity/agent-js/issues"
  },
  "keywords": [
    "internet computer",
    "internet-computer",
    "ic",
    "dfinity",
    "agent",
    "actor",
    "dfx",
    "canister",
    "candid",
    "motoko",
    "javascript",
    "typescript",
    "blockchain",
    "crypto",
    "distributed",
    "api",
    "sdk"
  ],
  "main": "./lib/cjs/index",
  "module": "./lib/esm/index",
  "unpkg": "./lib/esm/index",
  "scripts": {
    "build": "tsc -b && tsc -p tsconfig-cjs.json",
    "build:watch": "tsc -b --watch",
    "bundle": "esbuild --bundle src/index.ts --outfile=dist/index.js",
    "eslint:fix": "npm run lint -- --fix",
    "lint": "eslint --ext '.js,.jsx,.ts,.tsx' src/index.ts src/auth.ts src/request_id.ts src/request_id.test.ts",
    "lint:fix": "npm run lint -- --fix",
    "make:docs/reference": "typedoc src/index.ts --out ../../docs/generated/agent --excludeInternal",
    "release": "npm publish",
    "test": "jest --verbose",
    "test:coverage": "jest --verbose --collectCoverage"
  },
  "peerDependencies": {
    "@dfinity/candid": "^0.15.7",
    "@dfinity/principal": "^0.15.7"
  },
  "dependencies": {
    "base64-arraybuffer": "^0.2.0",
    "borc": "^2.1.1",
    "js-sha256": "0.9.0",
    "simple-cbor": "^0.4.1"
  },
  "devDependencies": {
    "@trust/webcrypto": "^0.9.2",
    "@types/jest": "^29.5.2",
    "@typescript-eslint/eslint-plugin": "^5.30.5",
    "@typescript-eslint/parser": "^5.30.5",
    "esbuild": "^0.15.16",
    "eslint": "^8.19.0",
    "eslint-plugin-jsdoc": "^39.3.3",
    "isomorphic-fetch": "^3.0.0",
    "jest": "^29.5.0",
    "jest-cli": "^29.5.0",
    "jest-environment-node": "^29.5.0",
    "size-limit": "^8.2.6",
    "text-encoding": "^0.7.0",
    "ts-jest": "^29.1.0",
    "ts-node": "^10.8.2",
    "typedoc": "^0.24.8",
    "typescript": "^5.1.5",
    "whatwg-fetch": "^3.0.0"
  },
  "size-limit": [
    {
      "path": "./dist/index.js",
      "limit": "100 kB",
      "webpack": false
    }
  ],
  "overrides": {
    "semver": "7.5.3",
    "word-wrap": "npm:@pavus/word-wrap@1.2.6"
  }
}
