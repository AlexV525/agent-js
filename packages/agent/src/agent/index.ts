export * from './http';
export * from './proxy';
