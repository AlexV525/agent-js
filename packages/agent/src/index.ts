export * from './actor.js';
export * from './agent/index.js';
export * from './auth.js';
export * from './certificate.js';
export * from './agent/http/transforms.js';
export * from './canisters/asset.js';
export * from './canisters/management.js';
export * from './fetch_candid.js';
export * from './request_id.js';
export * from './utils/bls.js';
export * from './utils/buffer.js';
export * as polling from './polling/index.js';
/**
 * The CanisterStatus utility is used to request structured data directly from the IC public API. This data can be accessed using agent.readState, but CanisterStatus provides a helpful abstraction with some known paths.
 *
 * You can request a canisters Controllers, ModuleHash, Candid interface, Subnet, or Time, or provide a custom path {@link CanisterStatus.CustomPath} and pass arbitrary buffers for valid paths identified in https://internetcomputer.org/docs/current/references/ic-interface-spec.
 *
 * The primary method for this namespace is {@link CanisterStatus.request}
 */
export * as CanisterStatus from './canisterStatus/index.js';

export * as Cbor from './cbor.js';
