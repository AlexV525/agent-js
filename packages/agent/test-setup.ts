// This file may be used to polyfill features that aren't available in the test
// environment, i.e. JSDom.
//
// We sometimes need to do this because our target browsers are expected to have
// a feature that JSDom doesn't.
//
// Note that we can use webpack configuration to make some features available to
// Node.js in a similar way.
import crypto from '@trust/webcrypto';
import fetch from 'isomorphic-fetch';
import { TextEncoder, TextDecoder } from 'text-encoding';
import { beforeEach } from 'vitest';

beforeEach(() => {
  global.TextEncoder = TextEncoder;
  global.TextDecoder = TextDecoder;
  global.crypto = crypto;
  global.fetch = fetch;
});
