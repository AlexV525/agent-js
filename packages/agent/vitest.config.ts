import { defineConfig } from 'vitest/config';

if (!globalThis.crypto) {
  global.crypto = require('@trust/webcrypto');
}

global.TextEncoder = require('text-encoding').TextEncoder;
global.TextDecoder = require('text-encoding').TextDecoder;
global.fetch = require('isomorphic-fetch');
export default defineConfig({});
