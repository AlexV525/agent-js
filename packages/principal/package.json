{
  "name": "@dfinity/principal",
  "version": "0.15.4",
  "author": "DFINITY Stiftung <sdk@dfinity.org>",
  "license": "Apache-2.0",
  "description": "JavaScript and TypeScript library to work with Internet Computer principals",
  "homepage": "https://internetcomputer.org",
  "repository": {
    "type": "git",
    "url": "https://github.com/dfinity/agent-js.git",
    "directory": "packages/principal"
  },
  "bugs": {
    "url": "https://github.com/dfinity/agent-js/issues"
  },
  "keywords": [
    "internet computer",
    "ic",
    "dfinity",
    "canister",
    "webauthn",
    "identity",
    "principal",
    "dfx",
    "candid",
    "motoko",
    "javascript",
    "typescript",
    "blockchain",
    "crypto",
    "distributed",
    "api"
  ],
  "type": "module",
  "main": "lib/esm/index.js",
  "exports": {
    ".": {
      "types": "./lib/esm/index.d.ts",
      "import": "./lib/esm/index.js",
      "require": "./lib/cjs/index.js"
    }
  },
  "scripts": {
    "build": "tsc -b && tsc -p tsconfig-cjs.json",
    "bundle": "esbuild --bundle src/index.ts --outfile=dist/index.js",
    "size-limit": "size-limit",
    "lint": "eslint 'src' --ext '.js,.jsx,.ts,.tsx'",
    "lint:fix": "npm run lint -- --fix",
    "make:docs/reference": "typedoc src/index.ts --out ../../docs/generated/principal --excludeInternal",
    "release": "npm publish",
    "test": "vitest run"
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^5.30.5",
    "@typescript-eslint/parser": "^5.30.5",
    "esbuild": "^0.15.16",
    "eslint": "^8.19.0",
    "eslint-plugin-jsdoc": "^39.3.3",
    "size-limit": "^8.1.0",
    "text-encoding": "^0.7.0",
    "ts-node": "^10.8.2",
    "tslint": "^5.20.0",
    "typedoc": "^0.22.11",
    "typescript": "^4.2.3",
    "vitest": "^0.29.3",
    "whatwg-fetch": "^3.0.0"
  },
  "dependencies": {
    "@dfinity/types": "^1.0.x",
    "js-sha256": "^0.9.0"
  },
  "size-limit": [
    {
      "path": "./dist/index.js",
      "limit": "100 kB",
      "webpack": false
    }
  ]
}
