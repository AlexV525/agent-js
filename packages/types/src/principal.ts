export abstract class AbstractPrincipal {
  readonly _isPrincipal = true;
  abstract isAnonymous(): boolean;
  abstract toUint8Array(): Uint8Array;
  abstract toHex(): string;
  abstract toText(): string;
  abstract toString(): string;
  /**
   * Utility method taking a Principal to compare against. Used for determining canister ranges in certificate verification
   * @param {Principal} other - a {@link Principal} to compare
   * @returns {'lt' | 'eq' | 'gt'} `'lt' | 'eq' | 'gt'` a string, representing less than, equal to, or greater than
   */
  abstract compareTo(other: AbstractPrincipal): 'lt' | 'eq' | 'gt';
  /**
   * Utility method checking whether a provided Principal is less than or equal to the current one using the {@link Principal.compareTo} method
   * @param other a {@link Principal} to compare
   * @returns {boolean} boolean
   */
  abstract ltEq(other: AbstractPrincipal): boolean;
  /**
   * Utility method checking whether a provided Principal is greater than or equal to the current one using the {@link Principal.compareTo} method
   * @param other a {@link Principal} to compare
   * @returns {boolean} boolean
   */
  abstract gtEq(other: AbstractPrincipal): boolean;
}
