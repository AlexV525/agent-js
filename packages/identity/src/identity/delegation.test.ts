import { SignIdentity } from '@dfinity/agent';
import { Principal } from '@dfinity/principal';
import { Delegation, DelegationChain } from './delegation';
import { Ed25519KeyIdentity } from './ed25519';

function createIdentity(seed: number): SignIdentity {
  const s = new Uint8Array([seed, ...new Array(31).fill(0)]);
  return Ed25519KeyIdentity.generate(s);
}

function randomPrincipal(): Principal {
  const bytes = new Uint8Array(29);
  crypto.getRandomValues(bytes);
  return Principal.fromUint8Array(bytes);
}

test('delegation signs with proper keys (3)', async () => {
  const root = createIdentity(2);
  const middle = createIdentity(1);
  const bottom = createIdentity(0);

  const rootToMiddle = await DelegationChain.create(
    root,
    middle.getPublicKey(),
    new Date(1609459200000),
  );
  const middleToBottom = await DelegationChain.create(
    middle,
    bottom.getPublicKey(),
    new Date(1609459200000),
    {
      previous: rootToMiddle,
    },
  );

  const golden = {
    delegations: [
      {
        delegation: {
          expiration: '1655f29d787c0000',
          pubkey:
            '302a300506032b6570032100cecc1507dc1ddd7295951c290888f095adb9044d1b73d696e6df065d683bd4fc',
        },
        signature:
          'b106d135e5ad7459dc67db68a4946fdbe603e650df4035957db7f0fb54e7467bb463116a2ad025e1887cd1f29025e0f3607b09924abbbbebfaf921b675c8ff08',
      },
      {
        delegation: {
          expiration: '1655f29d787c0000',
          pubkey:
            '302a300506032b65700321003b6a27bcceb6a42d62a3a8d02a6f0d73653215771de243a63ac048a18b59da29',
        },
        signature:
          '5e40f3d171e499a691092e5b961b5447921091bcf8c6409cb5641541f4dc1390f501c5dfb16b10df29d429cd153b9e396af4e883ed3cfa090d28e214db14c308',
      },
    ],
    publicKey:
      '302a300506032b65700321006b79c57e6a095239282c04818e96112f3f03a4001ba97a564c23852a3f1ea5fc',
  };

  expect(middleToBottom.toJSON()).toEqual(golden);
});

test('DelegationChain can be serialized to and from JSON', async () => {
  const root = createIdentity(2);
  const middle = createIdentity(1);
  const bottom = createIdentity(0);

  const rootToMiddle = await DelegationChain.create(
    root,
    middle.getPublicKey(),
    new Date(1609459200000),
    {
      targets: [Principal.fromText('jyi7r-7aaaa-aaaab-aaabq-cai')],
    },
  );
  const middleToBottom = await DelegationChain.create(
    middle,
    bottom.getPublicKey(),
    new Date(1609459200000),
    {
      previous: rootToMiddle,
      targets: [Principal.fromText('u76ha-lyaaa-aaaab-aacha-cai')],
    },
  );

  const rootToMiddleJson = JSON.stringify(rootToMiddle);
  // All strings in the JSON should be hex so it is clear how to decode this as different versions
  // of `toJSON` evolve.
  JSON.parse(rootToMiddleJson, (key, value) => {
    if (typeof value === 'string') {
      const byte = parseInt(value, 16);
      if (isNaN(byte)) {
        throw new Error(`expected all strings to be hex, but got: ${value}`);
      }
    }
    return value;
  });
  const rootToMiddleActual = DelegationChain.fromJSON(rootToMiddleJson);
  expect(rootToMiddleActual).toEqual(rootToMiddle);

  const middleToBottomJson = JSON.stringify(middleToBottom);
  const middleToBottomActual = DelegationChain.fromJSON(middleToBottomJson);
  expect(middleToBottomActual).toEqual(middleToBottom);
});

test('DelegationChain has a maximum length of 20 delegations', async () => {
  const identities = new Array(21).fill(0).map((_, i) => createIdentity(i));
  const signedDelegations: DelegationChain[] = [];

  for (let i = 0; i < identities.length - 1; i++) {
    const identity = identities[i];
    const nextIdentity = identities[i + 1];
    let signedDelegation: DelegationChain;
    if (i === 0) {
      signedDelegation = await DelegationChain.create(
        identity,
        nextIdentity.getPublicKey(),
        new Date(1609459200000),
      );
    } else {
      signedDelegation = await DelegationChain.create(
        identity,
        nextIdentity.getPublicKey(),
        new Date(1609459200000),
        {
          previous: signedDelegations[i - 1],
        },
      );
    }
    signedDelegations.push(signedDelegation);
  }
  expect(signedDelegations.length).toEqual(20);

  const secondLastIdentity = identities[identities.length - 2];
  const lastIdentity = identities[identities.length - 1];
  expect(
    DelegationChain.create(
      secondLastIdentity,
      lastIdentity.getPublicKey(),
      new Date(1609459200000),
      {
        previous: signedDelegations[signedDelegations.length - 1],
      },
    ),
  ).rejects.toThrow('Delegation chain cannot exceed 20');
});

test('Delegation can include multiple targets', async () => {
  const pubkey = new Uint8Array([1, 2, 3]);
  const expiration = BigInt(Number(new Date(1609459200000)));
  const targets = [
    Principal.fromText('jyi7r-7aaaa-aaaab-aaabq-cai'),
    Principal.fromText('u76ha-lyaaa-aaaab-aacha-cai'),
  ];
  const delegation = new Delegation(pubkey, expiration, targets);
  expect(delegation.targets).toEqual(targets);
  expect(delegation.toJSON()).toMatchSnapshot();
});

test('Delegation targets cannot exceed 1_000', () => {
  const targets = new Array(1_001).fill(randomPrincipal());
  expect(() => new Delegation(new Uint8Array([1, 2, 3]), BigInt(0), targets)).toThrow(
    'Delegation targets cannot exceed 1000',
  );
});

test('Delegation chains cannot repeat public keys', async () => {
  const root = createIdentity(0);
  const middle = createIdentity(1);
  const bottom = createIdentity(2);

  const rootToMiddle = await DelegationChain.create(
    root,
    middle.getPublicKey(),
    new Date(1609459200000),
  );
  const middleToBottom = await DelegationChain.create(
    middle,
    bottom.getPublicKey(),
    new Date(1609459200000),
    {
      previous: rootToMiddle,
    },
  );

  expect(
    DelegationChain.create(bottom, root.getPublicKey(), new Date(1609459200000), {
      previous: middleToBottom,
    }),
  ).rejects.toThrow('Delegation chain cannot repeat public keys');
});
