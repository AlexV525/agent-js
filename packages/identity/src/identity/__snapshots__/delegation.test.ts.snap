// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delegation can include multiple targets 1`] = `
Object {
  "expiration": "176bb3e7000",
  "pubkey": "010203",
  "targets": Array [
    "00000000002000030101",
    "000000000020008E0101",
  ],
}
`;
